src/Solana/Signer.cpp

```diff
@@ -0,0 +1,48 @@
+// Copyright © 2017-2019 Trust.
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+#include "Signer.h"
+#include <TrezorCrypto/ed25519.h>
+
+#include <algorithm>
+
+using namespace TW;
+using namespace TW::Solana;
+
+void Signer::sign(const std::vector<PrivateKey> &privateKeys, Transaction &transaction) {
+    for (auto privateKey : privateKeys) {
+        auto address = Address(privateKey.getPublicKey(TWPublicKeyTypeED25519));
+        auto index = transaction.getAccountIndex(address);
+        auto message = transaction.messageData();
+        auto signature = Signature(privateKey.sign(message, TWCurveED25519));
+        transaction.signatures[index] = signature;
+    }
+}
+
+void Signer::signUpdateBlockhash(const std::vector<PrivateKey> &privateKeys,
+                                 Transaction &transaction, Solana::Hash &recentBlockhash) {
+    transaction.message.recentBlockhash = recentBlockhash;
+    Signer::sign(privateKeys, transaction);
+}
+
+// This method does not confirm that PrivateKey order matches that encoded in the messageData
+// That order must be correct for the Transaction to succeed on Solana
```

Comment:
what does this mean?

Reply:
Solana allows for multiple signatures, and requires that signatures be provided in the same order as the corresponding public keys in the message body. The `Signer::sign()` method above examines the Transaction struct to place each signature at the correct index in the array.
The `Signer::signRawMessage()` method, in contrast, trusts the caller to pass private keys and place signatures in the correct order for the corresponding messageData.
This warning is moot for a single-signer transaction.